lisa/wlgen/workload.py

```diff
@@ -141,7 +139,42 @@ def wrap_command(self, cpus=None, cgroup=None):
             command = self.target.cgroups.run_into_cmd(cgroup, command)
 
         return command
+    def run_background(self, cpus=None, cgroup=None, as_root=False,
+                       stdout=PIPE, stderr=PIPE):
+        """
+        Execute the workload on the configured target asynchronously.
+
+        :param cpus: CPUs on which to restrict the workload execution (taskset)
+        :type cpus: list(int)
+
+        :param cgroup: cgroup in which to run the workload
+        :type cgroup: str
+
+        :param as_root: Whether to run the workload as root or not
+        :type as_root: bool
+
+        :returns: the Popen handle of the host subprocess executing the remote
+          command
+
+        .. note:: Popen handles can be used as context managers, so you could
+          (and probably should) use this like so::
+
+            with run_background() as handle:
+                do_stuff_with(handle)
+                handle.communicate()
+        """
+        logger = self.get_logger()
+        if not self.command:
```

Comment:
can this happen ? Maybe it would be better to catch that in `__init__` rather than here.

Reply:
Some workloads (e.g. sysbench) can have their command defined at `run()` time, which now that I look at it is pretty broken for splitting execution methods.

AFAICT I had done that to not have to recreate several `Sysbench` instances for slightly different invocations. I actually wrote a piece about that in the doc:

```
    ``__init__`` is there to initialize a given workload, and :meth:`run`                            
    can be called on it several times, with varying arguments.                                       
    As much work as possible should be delegated to :meth:`run`, so that                             
    different flavours of the same workload can be run without the hassle of                         
    creating a superfluous amount of new instances. However, when persistent                         
    data is involved (e.g. the workload depends on a file), then this data                           
    should be exposed as an ``__init__`` parameter.  
```

I need to have a think about this.

Reply:
The issue with that approach is that if we really do that, we will end up with everything as `run()` parameters and nothing in `__init__`. Anything in between will just be an arbitrary draw of whoever introduced the class, without any guarantee that the split will be sensible for the next user.
Functional programming comes to the rescue and can save us the `hassle of creating a superfluous amount of new instances` ! We put everything back in `__init__` and use currying:
```
import functools
class Workload:
     @classmethod
     def partial(cls, *args, **kwargs):
          return functools.partial(cls, *args, **kwargs)

class MyWorkload(Workload):
     def __init__(self, foo, bar):
         ....

make = MyWorkload.partial(foo=3)
make(bar=4).run()
make(bar=5).run() # No hassle !
```

To finish the work, we also need to have a way of amending the configuration after the workload has been created, for example for the RTA workloads (`run_rtapp()` may need to force some settings, e.g. disabling mlockall for CFS tasks when target is not rooted). This means pushing things to the target has to happen later than `__init__`. A nice place to store that is in `Workload.__enter__`, since it will be run right before we call `run()`. We can also have a "manual" `deploy()` method if somebody does not use the CM (or we can even do that automatically in `run()` if it has not been done already).

Reply:
Note that with a bit of metaprogramming we can even make the currying automatic, i.e.:
```
wload = Workload(foo=1)
wload(bar=3).run()
wload(bar=4).run()
# Raise an exception explicitly saying that "bar" was not given a value
wload.run()
```